##Major Project 3
Movie Recommendation system using one hot encoding.

Part 1: One hot encoding of the Director Column

In [1]:
import pandas as pd
import numpy as np

1) Loading the Dataset

In [5]:
df=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/final.csv')
print(df.head(1))

                       name  year  duration  genre  rating        director  \
0  The Shawshank Redemption  1994       142  Drama     9.3  Frank Darabont   

    type  
0  Movie  


2) Removing unnecessary Columns

In [6]:
#I will be recommending movies based on the director of the movie the user likes. 
#So I will remove every column except name and director.

del df['year']
del df['duration']
del df['genre']
del df['rating']
del df['type']

print(df.head())

                       name              director
0  The Shawshank Redemption        Frank Darabont
1             The Godfather  Francis Ford Coppola
2           The Dark Knight     Christopher Nolan
3    The Godfather: Part II  Francis Ford Coppola
4              12 Angry Men          Sidney Lumet


3) Checking for any null values

In [7]:
print(df.isnull().sum())
print('Number of rows: ', len(df))

name        0
director    0
dtype: int64
Number of rows:  1250


There are no null values and the total number of rows are 1250.

4) Finding all directors

In [8]:
directors=[]
for i in df['director']:
  directors+=i.split(',')

directors=[i.strip() for i in directors]
directors=list(set(directors))
print(len(set(directors)))

1210


5) One hot Conversion

In [9]:
one_hot=[]
for i in df['director']:
  i=[j.strip() for j in i.split(',')]
  enc=[0 for i in range(1210)]

  for direc in i:
    enc[directors.index(direc)]=1
  one_hot.append(enc)

6) One_hot verification

In [10]:
for i in range(len(df['director'])):
  x1=len(df['director'][i].split(','))
  x2=list(one_hot[i]).count(1)
  
  if(x1!=x2):
    print('Not Equal!')

One Hot Conversion for directors is successful!

7) Converting one_hot into array

In [11]:
one_hot=np.array(one_hot)
print(one_hot.shape)

(1250, 1210)


8) Adding one_hot data into the dataframe

In [12]:
for i in range(1210):
  df[directors[i]]=one_hot[:,i]

<ipython-input-12-7c9a838979e3>:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`
  df[directors[i]]=one_hot[:,i]


9) Recommending Movies (Top 10)

In [13]:
def mod(lst):
  s=0 
  for i in lst:
    if (i<0):
      s+=(-i)
    else:
      s+=(i)
  return s

In [14]:
user=input('Enter the choice of your movie: ')
user=df[df['name']==user].values[0][2:]

rel=[]
for i in df.values:
  rel.append(mod(i[2:]-user))

df['rel']=rel
df_1=df.sort_values(by='rel').head(10)

print(df_1['name'])

Enter the choice of your movie: The Godfather
1                   The Godfather
3          The Godfather: Part II
974       The Godfather: Part III
693              The Conversation
74                 Apocalypse Now
0        The Shawshank Redemption
671                    Awakenings
670                        Misery
669               Boyz n the Hood
668    La double vie de Véronique
Name: name, dtype: object


Above are the top 10 movies recommended based on the directors of the user's choice movie.

Part 2: One Hot Encoding Of the Movie name column.

1) Loading the dataset

In [74]:
mov=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/final.csv')
print(mov.head(1))

                       name  year  duration  genre  rating        director  \
0  The Shawshank Redemption  1994       142  Drama     9.3  Frank Darabont   

    type  
0  Movie  


2) Removing unnecessary columns

In [75]:
#One hot encoding has to be done only on the name column.
#Thus I will be removing all the other columns.

del mov['year']
del mov['duration']
del mov['genre']
del mov['rating']
del mov['director']
del mov['type']

print(mov.head())

                       name
0  The Shawshank Redemption
1             The Godfather
2           The Dark Knight
3    The Godfather: Part II
4              12 Angry Men


3) Removing ':' from the movie names as it affects one hot encoding.

In [76]:
temp=[]
temp_2=[]
name=[]
for i in mov['name']:
  temp.append(i)


for i in temp:
  if (':' in i[:]):
    temp_2.append(i.split(':'))
  else:
    name.append(i)

for i in temp_2:
  name.append(i[0]+i[1])

mov['name']=name

3) Checking for null values

In [77]:
print(mov['name'].isna().sum())
print(len(mov['name']))

0
1250


There are no null values. There are a total of 1250 rows.

4) Finding all Movie Names

In [102]:
temp=[]
for i in mov['name']:
  temp+=(i.split(' '))
#temp=list(set(temp))
print(len(temp))
#print(list(set(temp)))

3425


5) One hot coversion

In [109]:
one_hot_1=[]
for i in mov['name']:
  i=i.split(' ')
  enc=[0 for i in range(3425)]
  
  for nam in i:
    enc[temp.index(nam)]+=1

  one_hot_1.append(enc)

6) One hot verification

In [110]:
for j in range(len(mov['name'])):
  x1 = len(mov['name'][j].split(' '))
  x2 = 0
  for i in one_hot_1[j]:
    x2+=i
  if (x1!=x2):
   print(mov['name'][j])

One Hot Conversion of the Movie name is successful.

7) Converting One hot into an array

In [111]:
one_hot_1=np.array(one_hot_1)
print(one_hot_1.shape)

(1250, 3425)


8) Adding one_hot Data into the data frame

In [112]:
for i in range(2065):
  mov[temp[i]]=one_hot_1[:,i]

9) Recommending Top 10 movies based on Movie name

In [117]:
user=input('Enter the choice of your movie: ')
user=mov[mov['name']==user].values[0][1:]

rel = []
for i in mov.values:
  rel.append(mod(i[1:]-user))

mov['rel']=rel
#print(mov.head())
movi=mov.sort_values(by='rel', ascending=True).head(10)
print(movi['name'])

Enter the choice of your movie: The Godfather Part II
1162         The Godfather Part II
630     Back to the Future Part II
290          It Happened One Night
288             Gone with the Wind
666           To Have and Have Not
41              Back to the Future
133            There Will Be Blood
777              Der Name der Rose
370             Do the Right Thing
506       In the Heat of the Night
Name: name, dtype: object
